### PR TITLE
Upgrade from redux-persist v4 to redux-persist v5.

### PR DESCRIPTION
**Pull request by roberthoenig**
Please try this out and report how it works for you!

### CODE REVIEW

**Comment by roberthoenig** [comment]
(Flow will fail on this. From what I can see this is due to some issues with the redux-persist annotations.)

File: src/boot/store.js [hunk 1]
```diff
@@ -14,17 +15,20 @@ import middleware from './middleware';
 //   rootReducer,
 //   compose(autoRehydrate(), applyMiddleware(...middleware)),
 // );
+const reduxPersistConfigV4 = {
+  whitelist: [...config.storeKeys, ...config.cacheKeys],
```
Comment (borisyankov) [inline src/boot/store.js]: Extract to `const whitelist = [...config.storeKeys, ...config.cacheKeys]` and then reuse in both `reduxPersistConfigV4` and `reduxPersistConfig`.

File: src/boot/StoreProvider.js [hunk 2]
```diff
@@ -1,24 +1,25 @@
...
+import store, { persistor } from './store';
 import timing from '../utils/timing';
+import LoadingScreen from '../start/LoadingScreen';
 
 type Props = {
   children: ChildrenArray<*>,
 };
 
 export default class StoreHydrator extends PureComponent<Props> {
-  componentDidMount() {
```
Comment (borisyankov) [inline src/boot/StoreProvider.js]: We would want to keep this metric.
I guess we need to implement it differently.

Comment (borisyankov) [reply src/boot/StoreProvider.js]: Looking at this:
https://github.com/rt2zz/redux-persist/blob/c9ae1903c3c5dea8604dd79ceea50c8a0a64f8fd/src/integration/react.js#L38
`onBeforeLoad` is probably a good place to start the timing, but stopping the timing is not so clearly defined... zi would have hoped for a `onAfterLoad`.
We can do it on unmounting of the `loading` component.

File: src/diagnostics/Timer.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  }
+
+  componentWillUnmount() {
+    timing.end(this.props.name);
+  }
+
+  render() {
+    return this.props.children;
+  }
+}
```
Comment (borisyankov) [inline src/diagnostics/Timer.js]: Cool. That looks correct just looking at the code.
We might want to name this more descriptively. 'RenderTimer' sounds like an improvement. Any other alternatives you can come up with?

**Comment by roberthoenig** [comment]
Renamed Timer to RenderTimer.

**Comment by roberthoenig** [comment]
Updated to apply the compression first and then the upgrade to v5. **NOT TESTED YET**

**Comment by borisyankov** [comment]
Hydration on `master` on my S8 Plus takes 940msec on subsequent starts (not initially empty-store one)

Currently this PR is crashing on my phone, but will be debugging.

**Comment by roberthoenig** [comment]
Note that this PR is WIP, and I haven't pushed to it for a while so I'm expecting it to crash.

**Comment by roberthoenig** [comment]
Updated: Split into multiple commits.

File: src/boot/store.js [hunk 4]
```diff
@@ -15,17 +18,51 @@ import ZulipAsyncStorage from './ZulipAsyncStorage';
...
-      // $FlowFixMe: https://github.com/rt2zz/redux-persist/issues/823
-      storage: ZulipAsyncStorage,
-    },
-    onFinished,
-  );
+// DO NOT change this object. It contains the config for the last version
+// of the code that used redux-persist v4. This config object is used for
+// migrating any device that is ugprading from an app version that uses
+// redux-persist v4 to an app version that uses redux-persist v5.
+const reduxPersistConfigV4 = {
```
Comment (borisyankov) [inline src/boot/store.js]: You probably want to move this to a `migrations` file?

**Pull request closed without merging**